File: verify/iscostas.jl
# TODO: documentation


# Parsing the input
pint = Base.Fix1(parse, Int)

n = parse(Int, readline())
c = parse.(Int, split(readline(), " "))

function iscostas(perm::AbstractVector{<:Integer})::Bool
	n::Int = length(perm)
	length(unique(perm)) == n &&
	all(begin
		    seen::UInt64 = 0
		    all(begin
				diff::Int = perm[i + k] - perm[i] + n
				bit::UInt64 = UInt64(1) << diff
				result::Bool = seen & bit == 0
				seen |= bit
				result
			end for i in 1:(n - k))
	    end for k in 1:(n - 1))
end

println(iscostas(c) ? 1 : 0)

File: verify/iscostas.c
#include <stdio.h>
#include <string.h>

// TODO: documentation

int iscostas(int *perm, int n) {
    int seen[n + 1];
    memset(seen, 0, sizeof(seen));
    for (int i = 0; i < n; i++) {
        if (perm[i] < 1 || perm[i] > n || seen[perm[i]]) return 0;
        seen[perm[i]] = 1;
    }
    for (int k = 1; k < n; k++) {
        long long used = 0;
        for (int i = 0; i < n - k; i++) {
            int d = perm[i + k] - perm[i] + n;
            long long bit = 1LL << d;
            if (used & bit) return 0;
            used |= bit;
        }
    }
    return 1;
}

int main(void) {
    int n;
    (void)scanf("%d", &n);
    int perm[n];
    for (size_t i = 0; i < n; ++i) (void)scanf("%d", &perm[i]);
    printf("%d\n", iscostas(perm, n));
    return 0;
}
